rcl_logging_spdlog/src/rcl_logging_spdlog.cpp

```diff
@@ -128,6 +176,15 @@ rcl_logging_ret_t rcl_logging_external_initialize(
 
     auto sink = std::make_unique<spdlog::sinks::basic_file_sink_mt>(name_buffer, false);
     g_root_logger = std::make_unique<spdlog::logger>("root", std::move(sink));
+    if (!should_use_old_flushing_behavior) {
+      // in this case we should do the new thing (until config files are supported)
+      // which is to configure the logger to flush periodically and on
+      // error level messages
+      spdlog::flush_every(std::chrono::seconds(5));
```

Comment:
For what it is worth, this is not working like I expected.

I built rcl_logging with this branch on my Raspberry Pi.  Then I'm running a simple talker:

```
ros2 run demo_nodes_cpp talker
```

As expected, that prints out a message at log level INFO every second.  Also as expected, it creates a file in `~/.ros/log/talker_578699_1668519761895.log`.  However, it does *not* seem to be flushing every 5 seconds, as far as I can tell.  If I follow that log file with:

```
tail -F ~/.ros/log/talker_578699_1668519761895.log
```

It seems to be getting flushed ~ once a minute, though that may also have to do with how much data is being buffered up.

I should mention that my Raspberry Pi is still running Ubuntu 20.04, so has spdlog 1.5.0.

Can you double-check and see if that matches what you are seeing?  Maybe it is just a bug in spdlog, and maybe it is fixed in newer versions, but I want to check to be sure.

Reply:
Ok, I'll see if I can double check. What about the error message causing a sync?

I think there may be the chance for buffering between `tail -f` and the file as well. It might be better to check with something like `strace` that way we can see when the application is actually writing to the file.

Reply:
So, the periodic flushing is running, but it's not causing the file to be flushed:

```
william@markook-ubuntu-2204:~/ros2_ws$ strace ./install/demo_nodes_cpp/lib/demo_nodes_cpp/talker 2> >(grep 'write(')
write(2, "fake write(3: setting up flush_e"..., 39fake write(3: setting up flush_every()
write(7, "\1\0\0\0\0\0\0\0", 8)         = 8
write(13, "\1\0\0\0\0\0\0\0\200\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 32) = 32
write(15, "\1\0\0\0\0\0\0\0\200\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0", 32) = 32
write(2, "[INFO] [1668557105.393823678] [t"..., 69[INFO] [1668557105.393823678] [talker]: Publishing: 'Hello World: 1'
write(2, "[INFO] [1668557106.394955498] [t"..., 69[INFO] [1668557106.394955498] [talker]: Publishing: 'Hello World: 2'
write(2, "[INFO] [1668557107.393425570] [t"..., 69[INFO] [1668557107.393425570] [talker]: Publishing: 'Hello World: 3'
write(2, "[INFO] [1668557108.393684514] [t"..., 69[INFO] [1668557108.393684514] [talker]: Publishing: 'Hello World: 4'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101401, tv_nsec=826490442}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557109.393434667] [t"..., 69[INFO] [1668557109.393434667] [talker]: Publishing: 'Hello World: 5'
write(2, "[INFO] [1668557110.393749817] [t"..., 69[INFO] [1668557110.393749817] [talker]: Publishing: 'Hello World: 6'
write(2, "[INFO] [1668557111.394324509] [t"..., 69[INFO] [1668557111.394324509] [talker]: Publishing: 'Hello World: 7'
write(2, "[INFO] [1668557112.393437076] [t"..., 69[INFO] [1668557112.393437076] [talker]: Publishing: 'Hello World: 8'
write(2, "[INFO] [1668557113.394683890] [t"..., 69[INFO] [1668557113.394683890] [talker]: Publishing: 'Hello World: 9'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101406, tv_nsec=826489109}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557114.394135579] [t"..., 70[INFO] [1668557114.394135579] [talker]: Publishing: 'Hello World: 10'
write(2, "[INFO] [1668557115.395829849] [t"..., 70[INFO] [1668557115.395829849] [talker]: Publishing: 'Hello World: 11'
write(2, "[INFO] [1668557116.395629996] [t"..., 70[INFO] [1668557116.395629996] [talker]: Publishing: 'Hello World: 12'
write(2, "[INFO] [1668557117.396060099] [t"..., 70[INFO] [1668557117.396060099] [talker]: Publishing: 'Hello World: 13'
write(2, "[INFO] [1668557118.391452831] [t"..., 70[INFO] [1668557118.391452831] [talker]: Publishing: 'Hello World: 14'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101411, tv_nsec=826497025}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557119.394404096] [t"..., 70[INFO] [1668557119.394404096] [talker]: Publishing: 'Hello World: 15'
write(2, "[INFO] [1668557120.392766699] [t"..., 70[INFO] [1668557120.392766699] [talker]: Publishing: 'Hello World: 16'
write(2, "[INFO] [1668557121.392682424] [t"..., 70[INFO] [1668557121.392682424] [talker]: Publishing: 'Hello World: 17'
write(2, "[INFO] [1668557122.390958483] [t"..., 70[INFO] [1668557122.390958483] [talker]: Publishing: 'Hello World: 18'
write(2, "[INFO] [1668557123.392604204] [t"..., 70[INFO] [1668557123.392604204] [talker]: Publishing: 'Hello World: 19'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101416, tv_nsec=826498734}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557124.396831089] [t"..., 70[INFO] [1668557124.396831089] [talker]: Publishing: 'Hello World: 20'
write(2, "[INFO] [1668557125.390568692] [t"..., 70[INFO] [1668557125.390568692] [talker]: Publishing: 'Hello World: 21'
write(2, "[INFO] [1668557126.394710450] [t"..., 70[INFO] [1668557126.394710450] [talker]: Publishing: 'Hello World: 22'
write(2, "[INFO] [1668557127.396320294] [t"..., 70[INFO] [1668557127.396320294] [talker]: Publishing: 'Hello World: 23'
write(2, "[INFO] [1668557128.391827309] [t"..., 70[INFO] [1668557128.391827309] [talker]: Publishing: 'Hello World: 24'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101421, tv_nsec=826535608}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557129.392735651] [t"..., 70[INFO] [1668557129.392735651] [talker]: Publishing: 'Hello World: 25'
write(2, "[INFO] [1668557130.395698782] [t"..., 70[INFO] [1668557130.395698782] [talker]: Publishing: 'Hello World: 26'
write(2, "[INFO] [1668557131.392871418] [t"..., 70[INFO] [1668557131.392871418] [talker]: Publishing: 'Hello World: 27'
write(2, "[INFO] [1668557132.393328300] [t"..., 70[INFO] [1668557132.393328300] [talker]: Publishing: 'Hello World: 28'
write(2, "[INFO] [1668557133.396510303] [t"..., 70[INFO] [1668557133.396510303] [talker]: Publishing: 'Hello World: 29'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101426, tv_nsec=826499234}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557134.395815851] [t"..., 70[INFO] [1668557134.395815851] [talker]: Publishing: 'Hello World: 30'
write(2, "[INFO] [1668557135.396083398] [t"..., 70[INFO] [1668557135.396083398] [talker]: Publishing: 'Hello World: 31'
write(2, "[INFO] [1668557136.390593907] [t"..., 70[INFO] [1668557136.390593907] [talker]: Publishing: 'Hello World: 32'
write(2, "[INFO] [1668557137.396065363] [t"..., 70[INFO] [1668557137.396065363] [talker]: Publishing: 'Hello World: 33'
write(2, "[INFO] [1668557138.393292743] [t"..., 70[INFO] [1668557138.393292743] [talker]: Publishing: 'Hello World: 34'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101431, tv_nsec=826493734}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557139.392764454] [t"..., 70[INFO] [1668557139.392764454] [talker]: Publishing: 'Hello World: 35'
write(2, "[INFO] [1668557140.395244702] [t"..., 70[INFO] [1668557140.395244702] [talker]: Publishing: 'Hello World: 36'
write(2, "[INFO] [1668557141.394818702] [t"..., 70[INFO] [1668557141.394818702] [talker]: Publishing: 'Hello World: 37'
write(2, "[INFO] [1668557142.394319327] [t"..., 70[INFO] [1668557142.394319327] [talker]: Publishing: 'Hello World: 38'
write(2, "[INFO] [1668557143.392568410] [t"..., 70[INFO] [1668557143.392568410] [talker]: Publishing: 'Hello World: 39'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101436, tv_nsec=826489650}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557144.392757824] [t"..., 70[INFO] [1668557144.392757824] [talker]: Publishing: 'Hello World: 40'
write(2, "[INFO] [1668557145.395862609] [t"..., 70[INFO] [1668557145.395862609] [talker]: Publishing: 'Hello World: 41'
write(2, "[INFO] [1668557146.392036776] [t"..., 70[INFO] [1668557146.392036776] [talker]: Publishing: 'Hello World: 42'
write(2, "[INFO] [1668557147.396802974] [t"..., 70[INFO] [1668557147.396802974] [talker]: Publishing: 'Hello World: 43'
write(2, "[INFO] [1668557148.395905136] [t"..., 70[INFO] [1668557148.395905136] [talker]: Publishing: 'Hello World: 44'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101441, tv_nsec=829312064}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557149.399447334] [t"..., 70[INFO] [1668557149.399447334] [talker]: Publishing: 'Hello World: 45'
write(2, "[INFO] [1668557150.391735126] [t"..., 70[INFO] [1668557150.391735126] [talker]: Publishing: 'Hello World: 46'
write(2, "[INFO] [1668557151.390367578] [t"..., 70[INFO] [1668557151.390367578] [talker]: Publishing: 'Hello World: 47'
write(2, "[INFO] [1668557152.393246982] [t"..., 70[INFO] [1668557152.393246982] [talker]: Publishing: 'Hello World: 48'
write(2, "[INFO] [1668557153.398335509] [t"..., 70[INFO] [1668557153.398335509] [talker]: Publishing: 'Hello World: 49'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101446, tv_nsec=826490192}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557154.394059544] [t"..., 70[INFO] [1668557154.394059544] [talker]: Publishing: 'Hello World: 50'
write(2, "[INFO] [1668557155.394821574] [t"..., 70[INFO] [1668557155.394821574] [talker]: Publishing: 'Hello World: 51'
write(2, "[INFO] [1668557156.396664018] [t"..., 70[INFO] [1668557156.396664018] [talker]: Publishing: 'Hello World: 52'
write(2, "[INFO] [1668557157.393780591] [t"..., 70[INFO] [1668557157.393780591] [talker]: Publishing: 'Hello World: 53'
write(2, "[INFO] [1668557158.400238946] [t"..., 70[INFO] [1668557158.400238946] [talker]: Publishing: 'Hello World: 54'
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101451, tv_nsec=826488900}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557159.394116479] [t"..., 70[INFO] [1668557159.394116479] [talker]: Publishing: 'Hello World: 55'
write(2, "[INFO] [1668557160.396166586] [t"..., 70[INFO] [1668557160.396166586] [talker]: Publishing: 'Hello World: 56'
write(2, "[INFO] [1668557161.395948820] [t"..., 70[INFO] [1668557161.395948820] [talker]: Publishing: 'Hello World: 57'
write(2, "[INFO] [1668557162.393891931] [t"..., 70[INFO] [1668557162.393891931] [talker]: Publishing: 'Hello World: 58'
write(2, "[INFO] [1668557163.396294827] [t"..., 70[INFO] [1668557163.396294827] [talker]: Publishing: 'Hello World: 59'
write(3, "[INFO] [1668557105.393823678] [t"..., 4096) = 4096
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101456, tv_nsec=826489567}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
write(2, "[INFO] [1668557164.393267145] [t"..., 70[INFO] [1668557164.393267145] [talker]: Publishing: 'Hello World: 60'
write(2, "[INFO] [1668557165.392539043] [t"..., 70[INFO] [1668557165.392539043] [talker]: Publishing: 'Hello World: 61'
write(2, "[INFO] [1668557166.393528189] [t"..., 70[INFO] [1668557166.393528189] [talker]: Publishing: 'Hello World: 62'
^C
```

I added some code to the flushing callback in spdlog, which is where these lines are coming from:

```
futex(0xaaaad137b238, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=101456, tv_nsec=826489567}, FUTEX_BITSET_MATCH_ANYfake write(3: flushing
```

You can see that after accumulating 4096 bytes it writes to the file in this line:

```
write(3, "[INFO] [1668557105.393823678] [t"..., 4096) = 4096
```

So, it looks like the flush isn't going through. The docs (https://spdlog.docsforge.com/v1.x/7.flush-policy/) clearly say it only works if the sink is thread-safe, but we're using the `_mt` file sink, so it should work... I'll keep digging.

Reply:
Ok, so it ends up being that we were not properly registering our logger with spdlog, see: [57942c6](https://github.com/ros2/rcl_logging/pull/95/commits/57942c6ffa745f3b141f5cdf615080d6eafac9cc)

I think this is the right thing to do, in general (whether or not we use periodic flushing) and we missed it only because we're initializing the logger in a bit of a strange way. Normally it's not required, but since we're "manually" creating our logger we need to register it manually too: https://spdlog.docsforge.com/v1.x/5.logger-registry/#registering-new-loggers

spdlog keeps `shared_ptr`s to its registered loggers, so we have to store it as a `shard_ptr` (instead of `unique_ptr` too), but I don't think that will be an issue. I also drop (unregister) the logger as well in the shutdown.

Reply:
Thanks.  With this in place, this actually does what I expect now; it is flushing every 5 seconds.

I'm going to add a longer comment down below with the results of some testing I did.